core/state/snapshot/generate.go

```diff
@@ -159,169 +179,619 @@ func journalProgress(db ethdb.KeyValueWriter, marker []byte, stats *generatorSta
 	rawdb.WriteSnapshotGenerator(db, blob)
 }
 
-// generate is a background thread that iterates over the state and storage tries,
-// constructing the state snapshot. All the arguments are purely for statistics
-// gathering and logging, since the method surfs the blocks as they arrive, often
-// being restarted.
-func (dl *diskLayer) generate(stats *generatorStats) {
-	// If a database wipe is in operation, wait until it's done
-	if stats.wiping != nil {
-		stats.Log("Wiper running, state snapshotting paused", common.Hash{}, dl.genMarker)
-		select {
-		// If wiper is done, resume normal mode of operation
-		case <-stats.wiping:
-			stats.wiping = nil
-			stats.start = time.Now()
-
-		// If generator was aborted during wipe, return
-		case abort := <-dl.genAbort:
-			abort <- stats
-			return
+// proofResult is returned by the range provers
+type proofResult struct {
+	// These fields are set if the verification failed.
+	staleKeys   [][]byte // The stale keys to be deleted or overwritten (nil if verification passed).
+	staleValues [][]byte // The stale values to be deleted or overwritten (nil if verification passed).
+
+	// These fields are used/set if the verification passed
+	continued bool   // true if there are more elements in this trie.
+	last      []byte // the last key (if check was successfull).
+	count     int    // number of elements (set if verification passed)
+
+	tr *trie.Trie // trie for the given root, which can be used by the caller afterwards (may be nil)
+}
+
+// proveStorageRange checks whether the snapshot storage range is already complete, and if not,
+// recovers the data using the trie.
+func (dl *diskLayer) proveStorageRange(root common.Hash, accountHash common.Hash, origin []byte, max int) (*proofResult, error) {
+	var (
+		keys    = make([][]byte, 0) // Init to non-nil, to differentiate against nil returnvalue which means OK
+		vals    = make([][]byte, 0)
+		prefix  = append(rawdb.SnapshotStoragePrefix, accountHash.Bytes()...)
+		iter    = dl.diskdb.NewIterator(prefix, origin)
+		aborted bool
+		last    []byte
+	)
+	for iter.Next() {
+		key := iter.Key()
+		if len(key) != len(prefix)+common.HashLength {
+			continue // random state with same prefix
 		}
+		if len(keys) == max {
+			// There exists more data in the snapshot, but we've reached the
+			// max elems per iteration
+			aborted = true
+			break
+		}
+		keys = append(keys, common.CopyBytes(key[len(prefix):]))
+		vals = append(vals, common.CopyBytes(iter.Value()))
+	}
+	iter.Release()
+	if len(keys) > 0 {
+		last = keys[len(keys)-1]
 	}
-	// Create an account and state iterator pointing to the current generator marker
-	accTrie, err := trie.NewSecure(dl.root, dl.triedb)
+	// If the elements cover the entire storage range, then we don't need to use the prover
+	// to check validity. We can just feed the key/vals into a stacktrie and obtain the
+	// root hash
+	if len(origin) == 0 && !aborted {
+		stackTr := trie.NewStackTrie(nil)
```

Comment:
Actually in this case, we don't need the stacktrie. For tiny storage trie, we don't need to build the edge proofs at all. We can pass the nil proof and the entire key/val set to the range prover. 

Regarding the standard trie VS stacktrie, I think there is no big difference in the performance. StackTrie is more memory friendly, but for the tiny trie it's unnecessary to introduce the stacktrie.